### DISCUSSION TITLE
Missing submodule command - git submodule update --recursive --remote

### DISCUSSION THREAD

**Discussion by laurinkeithdavis** [gitextensions/gitextensions#11509] [Q&A] [answered]
Am I crazy or is this missing as an option?

`git submodule update --recursive --remote`

I don't understand why "Update all submodules" does this instead?

`git submodule update --init --recursive`

That command does not pull from the remote.

**Comment by laurinkeithdavis** [comment]
FYI,  know I can do that with a script and I have, but seems like this is a common use, while --init is only for when a repo is first checked out.

**Reply by vbjay** [reply]
Nope. https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt-update--init--remote-N--no-fetch--no-recommend-shallow-f--force--checkout--rebase--merge--referenceltrepositorygt--depthltdepthgt--recursive--jobsltngt--no-single-branch--filterltfilterspecgt--ltpathgt82308203

Update init is exactly what you want to say I want to make submodule be at point of where parent expects. It is up to developer to commit a change in parent repo saying it points to new commit.

**Reply by vbjay** [reply]
Also init is ok. It makes sure submodule is setup. It is not harmful if ran in later updates. 

https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt---init

**Reply by laurinkeithdavis** [reply]
But it does not work, these are not the same commands and do not do the same thing. When I do "Update all submodules" (which is `git submodule update --recursive --remote`), it does not update the submodules to the most recent commit in the branch they are in, but if I do `git submodule update --recursive --remote` from the command line it does. I've tested this multiple times. "Update all submodules" appears to be pulling down the changes, but when I switch to that submodule, the new commits are not there and the current revision has not changed. 

Not knowing these commands terribly well, maybe the issue is that the "Update all submodules" command does not include `--remote`. Yea, I bet that is what I am trying to get at (sorry, was confused about `--init`):

https://git-scm.com/docs/git-submodule#Documentation/git-submodule.txt---remote

I'll test this now.

**Reply by vbjay** [reply]
That's because update all submodules isnt set to do that. It is set to make sure that when you move your parent repo HEAD that the sub modules sync up to be where that commit points in the sub module.

**Reply by laurinkeithdavis** [reply]
Well, but that's what I am saying, the command we need is missing. I don't even see how that is useful doing an update without `--remote` is useful, as that means I have to switch to each submodule and fetch to update all that repo to the current remote head. 

The current command is definitely not useful to us, as we want all submodules to stay current with their remote head. That's why this discussion was titled "missing command". :)

Can you explain the purpose of this?

> It is set to make sure that when you move your parent repo HEAD that the sub modules sync up to be where that commit points in the sub module.

How do I set or change the "parent repo HEAD" for a submodule? I guess that has to be done manually?

**Reply by vbjay** [reply]
Making this change
![image](https://github.com/gitextensions/gitextensions/assets/3629489/56e4f14d-546f-48a1-908b-ed4b7f8ae377)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/cc222766-888e-4ce9-a738-5afeb03ff92e)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/f8bcbf30-3fe0-4c80-a80d-72d90bf1ace4)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/03b0705a-2799-4d84-955d-1de511e84c82)

Parent repo already staged 
![image](https://github.com/gitextensions/gitextensions/assets/3629489/55658a1a-2beb-450f-a308-97425cb8f6ad)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/dc12c1f0-49ed-472a-b014-0b4c6a3187c3)

So now I have a just a test commit that has a sub module change.  

Let's create a branch here 
![image](https://github.com/gitextensions/gitextensions/assets/3629489/7a12b9ac-1347-49c9-a3be-f48d73120247)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/5506a796-fa62-4392-a251-ef8cad67adc3)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/c36742cd-d848-4360-a761-a01dab6033b2)

Do you see how update submodule runs and makes sure all my submodules are at x point based on the commit in the GE repo?  That command does fetch missing commits if needed as seen here  
![image](https://github.com/gitextensions/gitextensions/assets/3629489/007aed9a-c978-4de6-b360-272217254f6d)
Niow in the submodule I check out 
![image](https://github.com/gitextensions/gitextensions/assets/3629489/3c1e92d2-ec5b-4dfb-b3d2-ef16f2a4a317)
and then go to GE repo and... 
![image](https://github.com/gitextensions/gitextensions/assets/3629489/91d0f196-037d-4710-833a-025dce151c30)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/dba10ae2-4968-40df-ac6e-a455c00649bf)

If you want to do stuff in the submodule like check out latestfrom x branch.  Do so the parent repo will reflect that status but you manage the submodules like so.  
![image](https://github.com/gitextensions/gitextensions/assets/3629489/5d530374-3563-4dbe-a8dc-fe691519348a)
![image](https://github.com/gitextensions/gitextensions/assets/3629489/1050787a-cc33-4710-b47a-039539702c16)

If you want to sync up changes in a submodule with latest then do a fetch/pull in that submodule. Update all submodules is geared to making sure you can start your work with all submodules pointed to a specific commit when you bounce like when I check out master and it moves the submodule back. to my just testing commit 
![image](https://github.com/gitextensions/gitextensions/assets/3629489/2b7085db-d802-4195-a3a6-8459787fe26f)

**Reply by vbjay** [reply]
As long as you follow master being the correct branch  to sync off... this script works
```bash
-c "git submodule foreach git checkout master && 
git submodule foreach git pull upstream master && 
git submodule foreach git push origin master && 
git checkout master && 
git pull upstream master && 
git push origin master &&
git submodule update --init --recursive"
```
![image](https://github.com/gitextensions/gitextensions/assets/3629489/6ee84c7e-3b7d-438b-8b14-1a5b7be14ac2)

**Comment by laurinkeithdavis** [comment]
Wow, thanks for all that info, that must have taken you a while.

> If you want to sync up changes in a submodule with latest then do a fetch/pull in that submodule.

Ok, that's what I thought you were getting at, and yes, we don't want to do that, we want the submodule to stay current with the head of its remote. 

> If you want to do stuff in the submodule like check out latestfrom x branch. Do so the parent repo will reflect that status but you manage the submodules like so.

Tried that, but still does not fetch from the remote. Maybe I'm missing something here.

> As long as you follow master being the correct branch to sync off... this script works

Not sure why we would need to do all of that, as `git submodule update --recursive --remote` as the script seems to do all that we need. We don't need the local master of the submodules to be updated, we just need the checked-out revision to be the most recent on the remote. Anyway, I'll try your script and compare the difference. Again, thanks.

**Reply by vbjay** [reply, marked as answer]
I did that to get my master in my forks synced up to my remotes so yes it does more.  If you want to add the pr that adds the menu option, then some points:

-  Synchronize menu does something else and is not what you want.  
- I would ask @RussKie  what the menu should say.
- The complexity issue with a menu option that does what you ask is that what if git config is not pointed to a specific branch as the source of truth to sync latest changes?  Does it just do a defacto master or what?

**Reply by vbjay** [reply]
For now I would add a script that does that and play with it.  I would also create a submodule that doesn't have branch filled in and set up the remotes and see what that does.  Note the research in your pr showing the menu behaves like this in these states.

**Comment by eng-marani** [comment]
I had the same issue. I know is not a typical workflow, but sometimes you need a submodule to always point to the "head" revision of the selected branch as configured, and there is not an asy way other than use a git script or go manually on each submodule one by one fetching/merging from the head, than sync the main repository again recursively

**Reply by vbjay** [reply]
That's correct. As it should be. It would be a nightmare if it was the other way. If I move my head in my super repo to a commit, as expected I expect the sub repo to be checked out at the commit the super repo points to. Look at the .git file you commit when you do move the submodule to a new commit. It is just a sha. We on purpose do not have the functionality you are asking for. You can develop a script like mine. Mine does more than just sync up from remote and I could have used inputs in the script to select remotes and branch names. 

What happens when you use your command you want to run and the submodule is in a headless state? Which is most of the time after being put in the right state after update init.

If you want the command add it to scripts and test it out. Test with both headless state and a branch checked out.

**Reply by eng-marani** [reply]
The request is, in submodule configuration, to have some extra flag to always sync to the head (maintaining the currently selected branch) so no script is really necessary anymore. Of course this should be activated on some specific submodules, not everywhere recursively. Nothing wrong to have this as "opt-in" and disabled by default.
